File: py/graph and trees/formulas_in_propositional_calculus.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

def is_operator(c):
    return c in ['~', '&', '|', '>', '=']

def create_tree(expr):
    stack = []
    for c in expr:
        if c == ' ':
            continue
        elif c.isalpha():  # operand
            stack.append(Node(c))
        elif is_operator(c):
            node = Node(c)
            if c == '~':  # unary operator
                node.right = stack.pop()
            else:  # binary operator
                right = stack.pop()
                left = stack.pop()
                node.right = right
                node.left = left
            stack.append(node)
    return stack[-1]  # root node

# Postorder traversal to verify
def postorder(root):
    if not root:
        return
    postorder(root.left)
    postorder(root.right)
    print(root.value, end=' ')

# Example usage
expr = "PQ&R|"   # equivalent to (P & Q) | R in postfix form
root = create_tree(expr)
print("Postorder traversal of tree:")
postorder(root)

File: ds practical/cpp/graph and tree/formulas_in_propositional_calculus.cpp
#include <iostream>
#include <stack>
using namespace std;

struct Node {
    char value;
    Node *left, *right;
    Node(char v) : value(v), left(NULL), right(NULL) {}
};

bool isOperator(char c) {
    return c == '~' || c == '&' || c == '|' || c == '>' || c == '=';
}

Node* createTree(string expr) {
    stack<Node*> st;
    for (char c : expr) {
        if (c == ' ') continue;
        if (isalpha(c)) st.push(new Node(c));
        else if (isOperator(c)) {
            Node* node = new Node(c);
            if (c == '~') {
                node->right = st.top(); st.pop();
            } else {
                Node* right = st.top(); st.pop();
                Node* left = st.top(); st.pop();
                node->left = left;
                node->right = right;
            }
            st.push(node);
        }
    }
    return st.top();
}

void postorder(Node* root) {
    if (!root) return;
    postorder(root->left);
    postorder(root->right);
    cout << root->value << " ";
}

int main() {
    string expr = "PQ&R|"; // postfix form
    Node* root = createTree(expr);
    cout << "Postorder traversal:\n";
    postorder(root);
    cout << endl;
}
